### ISSUE TITLE
Do accessibility audit (screen readers) of the site

### ISSUE THREAD

**Issue by joelwass** [teacherfund/TeacherFund_next#106]
Make sure our site meets a11y standards (https://www.w3.org/WAI/standards-guidelines/wcag/)

**Comment by mjfneto** [comment]
Hi, can I give it a try?

**Comment by stripedpajamas** [comment]
Absolutely @mjfneto

**Comment by mjfneto** [comment]
@stripedpajamas After some research, I think someone more experienced in Next.js could take the lead of this issue. Since I'm still learning about the framework and I don't know if you have a target date for this step, it might be better to have the help of a more proficient contributer.

- I have been following [WHUCAG 2.0 checklists](https://www.wuhcag.com/wcag-checklist/) if someone is interested in giving a hand. 

- Currently looking for more info on [`Custom <Document>`](https://nextjs.org/docs#custom-document) to set the `lang` attribute of the `html` element.
    - [Is there an easier way to define a language for the page? zeit/next.js #2256](https://github.com/zeit/next.js/issues/2256)
    - [What is the best way to set language attribute on <html>?](https://spectrum.chat/next-js/general/what-is-the-best-way-to-set-language-attribute-on-html~73aff546-0f5e-4f01-9ced-f45d4fb53f35)

**Comment by N7Layla** [comment]
Still looking for someone to work on this? This seems like a really valuable project so I'd love to help out, especially on the accessibility front. I ran the site through the axe tool and found these - I'd love to try and submit a fix for at least a few of them:
![image](https://user-images.githubusercontent.com/19983454/65984968-e0b43800-e44e-11e9-98c7-1b5e29dfa2b3.png)

**Comment by joelwass** [comment]
That would be incredible! Thank you @N7Layla

**Comment by N7Layla** [comment]
Just submitted a PR for several of the fixes, unfortunately I had a lot of trouble getting the Document setup working to add the lang attribute, so here are the major accessibility violations remaining:
 - HTML must have lang attribute (see above information - needs a _document file but I think it may need something additional to play nicely with the _app file)
 - Elements must have sufficient color contrast (all of the buttons on the main page fall into this, but not sure if you're wanting to make changes to design specs at this time)

**Comment by danomah** [comment]
Hey Joel @joelwass, I can run a few deep tests with a few screen readers if you are still looking for help? Also on the AXE Pro Beta programme and I know they are looking for as many tests to be run using the tool. Should produce some good results.

**Comment by joelwass** [comment]
That would be amazing! @danomah

**Comment by danomah** [comment]
Okily dokily. i'm not the best at react (as in running it in the browser). I've installed NPM and react. When I cd into the folder and run npm start i'm getting nowhere. I know it's something I'm doing wrong. Any advice?

**Comment by DustinHaefele** [comment]
@danomah are you running npm run build first? With next you have to build before starting.

**Comment by danomah** [comment]
I shall try that. Sorry it's been a while since I've used React. Thanks
Dustin
D

**Comment by stripedpajamas** [comment]
Have you read through the contributing guide linked from the readme? The steps to get up and running should roughly be:

- make sure you have node and npm installed (node v8 and higher)
- clone the repo, cd into the folder
- run `npm install`
- run `npm run dev`

The last command should build and watch for changes. You should then be able to see the app at http://localhost:3001

@danomah @DustinHaefele

**Comment by danomah** [comment]
@stripedpajamas @DustinHaefele That would have been clever of me. :) Thanks Peter and Dustin. Will have a look at it over the weekend. Appreciate it guys.
Dan

**Comment by meowwwls** [comment]
I would like to take some time going through the full site, but for now, I found several important issues with the donation form, arguably the most important element of the site.

- Even though you are using `aria-label` attributes on form elements to let screenreader users know the purpose of each input, every use benefits from visible labels, especially on a form with multiple inputs (rather than, say, a newsletter signup where email is the only thing requested). If I were filling out this form and got distracted at some point by a phone call or something else, once I've begun typing anything in the field, it is easy to forget what that field is for. This also makes it harder to recheck information before submitting. I suggest visible labels for every form element and ditching the `aria-label` which only helps some users. Blindness is not the only accessibility concern. 

- The input fields in the donation form would benefit from more space between them, especially for mobile users. It can be easy to tap the wrong field, especially without much contrast between the field and the background.

- The donate button needs to be an actual button element. Right now you are assigning a click event to a div which is not only not semantic, but you lose the built-in benefits of using a button which include keyboard accessibility and click / keyboard input handling. Right now, a keyboard user has no way to submit the form.

- label elements need an associated form input, but here you're using them for the `donate` text and the 'give once' and 'monthly' text. 

- The donation frequency options should be radio buttons wrapped in a fieldset so that not only are they keyboard accessible, but a screenreader user knows what the options are referring to.

- If the field is required, you should be using the `aria-required` attribute to let the screenreader user know.

- Right now if I try to submit without filling out any information, the only error I get is that the credit card information is invalid. Each invalid form field should give a message (and then go away once the field becomes valid) and use the `aria-invalid` attribute until the field is validated. 

I would like to work on some of these issues (today I have time to handle the `aria-required` and donate button issue), but can offer guidance if needed for anyone wanting to work on other issues! This is a wonderful project and I would like to help in any way.

**Comment by joelwass** [comment]
This is awesome, thanks @meowwwls !

**Comment by meowwwls** [comment]
> * HTML must have lang attribute (see above information - needs a _document file but I think it may need something additional to play nicely with the _app file)

@N7Layla I was banging my head against the wall for a couple hours over this because I was following the docs and was getting server errors. Apparently, the documentation's example for a custom document page is for current versions of Next and this project is using ~v7. What was causing the error was they were importing and using `Html`, which is new to v8 or v9. Removing that and using a regular `html` element solved the issue.

**Comment by joelwass** [comment]
Should we open an issue to upgrade to latest Next version? @meowwwls

**Comment by meowwwls** [comment]
@joelwass I think it's a good idea! I'm not sure if there are any breaking changes and I've never worked with Next until this project, so I definitely think a separate issue where those with more experience can talk it out is the way to go.

**Comment by joelwass** [comment]
This issue has been sufficiently addressed, thanks you to @meowwwls for taking the lead on accessibility! closing issue for now

**Comment by meowwwls** [comment]
@joelwass Would it be helpful to open independent issues for the remaining accessibility problems? They almost all involve design decisions as well, so I do think each having their own issue would be best so there can be a discussion. (Color contrast issues, lack of visible labels on form elements, and the mobile menu are 3 of the biggest remaining issues off the top of my head.)

**Comment by joelwass** [comment]
Definitely would be useful - isolating them also makes them more easily addressed and not too broad as well, thanks @meowwwls  :)
